### PR TITLE
feature/FOUR-9623: The default needs to list first in the login page

### PR DESCRIPTION
**Pull request by luNunezProcessmaker**
## Issue & Reproduction Steps
When the final user go to login in ProcessMaker, will be redirect to the selected provider.

## Related Tickets & Packages
- https://processmaker.atlassian.net/browse/FOUR-9623

## Code Review Checklist
- [ ] I have pulled this code locally and tested it on my instance, along with any associated packages.
- [ ] This code adheres to [ProcessMaker Coding Guidelines](https://github.com/ProcessMaker/processmaker/wiki/Coding-Guidelines).
- [ ] This code includes a unit test or an E2E test that tests its functionality, or is covered by an existing test.
- [ ] This solution fixes the bug reported in the original ticket.
- [ ] This solution does not alter the expected output of a component in a way that would break existing Processes.
- [ ] This solution does not implement any breaking changes that would invalidate documentation or cause existing Processes to fail.
- [ ] This solution has been tested with enterprise packages that rely on its functionality and does not introduce bugs in those packages.
- [ ] This code does not duplicate functionality that already exists in the framework or in ProcessMaker.
- [ ] This ticket conforms to the PRD associated with this part of ProcessMaker.

ci:next

### CODE REVIEW

File: ProcessMaker/Http/Controllers/Auth/LoginController.php [hunk 1]
```diff
@@ -55,6 +56,19 @@ public function showLoginForm()
...
         $addons = $manager->list();
+        $default = Setting::where('key', 'sso.default.login')->first();
+
+        // Redirectind to default login
+        if (!empty($default) && !empty($addons->toArray())) {
+            $position = $default->getAttribute('config');
+            $element = $default->getAttribute('ui')->elements[$position];
+            $drivers = $addons->toArray()[0]->data['drivers'];
+            
+            if ( array_key_exists(strtolower($element->name), $drivers)) {
```
Comment (fagubla) [inline ProcessMaker/Http/Controllers/Auth/LoginController.php:67]: remove extra space

**Review by fagubla — CHANGES_REQUESTED** [review]

File: ProcessMaker/Http/Controllers/Auth/LoginController.php [hunk 2]
```diff
@@ -55,6 +56,19 @@ public function showLoginForm()
     {
         $manager = App::make(LoginManager::class);
         $addons = $manager->list();
+        $default = Setting::where('key', 'sso.default.login')->first();
+
+        // Redirectind to default login
+        if (!empty($default) && !empty($addons->toArray())) {
+            $position = $default->getAttribute('config');
+            $element = $default->getAttribute('ui')->elements[$position];
+            $drivers = $addons->toArray()[0]->data['drivers'];
```
Comment (pmPaulis) [inline ProcessMaker/Http/Controllers/Auth/LoginController.php:65]: Use laravel helper to get the head of array https://laravel.com/docs/10.x/helpers

**Review by pmPaulis — CHANGES_REQUESTED** [review]

File: ProcessMaker/Http/Controllers/Auth/LoginController.php [hunk 3]
```diff
@@ -55,6 +56,18 @@ public function showLoginForm()
     {
         $manager = App::make(LoginManager::class);
         $addons = $manager->list();
+        $default = Setting::where('key', 'sso.default.login')->first();
+
+        // Redirectind to default login
+        if (!empty($default) && !empty($addons->toArray())) {
+            $position = $default->getAttribute('config');
+            $element = $default->getAttribute('ui')->elements[$position];
+            $drivers = head($addons->toArray())->data['drivers'];
```
Comment (pmPaulis) [inline ProcessMaker/Http/Controllers/Auth/LoginController.php:65]: What happens if the data['drivers'] does not exist?
Please validate this

**Review by pmPaulis — CHANGES_REQUESTED** [review]

File: ProcessMaker/Http/Controllers/Auth/LoginController.php [hunk 4]
```diff
@@ -55,6 +56,18 @@ public function showLoginForm()
     {
         $manager = App::make(LoginManager::class);
         $addons = $manager->list();
+        $default = Setting::where('key', 'sso.default.login')->first();
+
+        // Redirectind to default login
+        if (!empty($default) && !empty($addons->toArray())) {
+            $position = $default->getAttribute('config');
+            $element = $default->getAttribute('ui')->elements[$position];
```
Comment (pmPaulis) [inline ProcessMaker/Http/Controllers/Auth/LoginController.php:64]: What happens if the elements[$position] does not exist?
Please validate this

**Review by pmPaulis — CHANGES_REQUESTED** [review]

File: ProcessMaker/Http/Controllers/Auth/LoginController.php [hunk 5]
```diff
@@ -55,6 +56,25 @@ public function showLoginForm()
     {
         $manager = App::make(LoginManager::class);
         $addons = $manager->list();
+        $default = Setting::where('key', 'sso.default.login')->first();
```
Comment (gproly) [inline ProcessMaker/Http/Controllers/Auth/LoginController.php:59]: Is it possible to obtain $element->name from a database query?

**Review by gproly — CHANGES_REQUESTED** [review]

File: ProcessMaker/Http/Controllers/Auth/LoginController.php (hunk 5 above)
Comment (luNunezProcessmaker) [reply ProcessMaker/Http/Controllers/Auth/LoginController.php:61]: As i show you this element handles a fairly extensive json in which the name is found within it, so it is not possible

File: ProcessMaker/Http/Controllers/Auth/LoginController.php [hunk 6]
```diff
@@ -55,6 +58,25 @@ public function showLoginForm()
...
+            if (isset($position) && !empty($drivers)) {
+                $elements = $default->getAttribute('ui')->elements;
+                if (count($elements) >= $position + 1) {
+                    $element = $elements[$position];
+                    if (array_key_exists(strtolower($element->name), $drivers)) {
+                        return redirect()->route('sso.redirect', ['driver' => strtolower($element->name)]);
+                    }
+                }
+            }
+
```
Comment (pmPaulis) [inline ProcessMaker/Http/Controllers/Auth/LoginController.php:77]: remove this extra line

File: ProcessMaker/Http/Controllers/Auth/LoginController.php [hunk 7]
```diff
@@ -55,6 +58,25 @@ public function showLoginForm()
     {
         $manager = App::make(LoginManager::class);
         $addons = $manager->list();
+        $default = Setting::where('key', 'sso.default.login')->first();
+
+        // Redirectind to default login
+        if (!empty($default) && !empty($addons->toArray())) {
+            $position = $default->getAttribute('config'); // Type int
+            $data = head($addons->toArray())->data;
```
Comment (pmPaulis) [inline ProcessMaker/Http/Controllers/Auth/LoginController.php:66]: $addons->toArray() is converted twice
please validate if **data** property exist

**Review by pmPaulis — CHANGES_REQUESTED** [review]

File: ProcessMaker/Http/Controllers/Auth/LoginController.php [hunk 8]
```diff
@@ -10,9 +10,12 @@
 use ProcessMaker\Events\Logout;
 use ProcessMaker\Http\Controllers\Controller;
 use ProcessMaker\Managers\LoginManager;
+use ProcessMaker\Models\Setting;
 use ProcessMaker\Models\User;
 use ProcessMaker\Traits\HasControllerAddons;
 
+use function Widmogrod\Functional\length;
```
Comment (pmPaulis) [inline ProcessMaker/Http/Controllers/Auth/LoginController.php:17]: use function ?? Remove this unused line

**Review by pmPaulis — CHANGES_REQUESTED** [review]

**Review by pmPaulis — APPROVED** [review]

**Pull request merged**